includes/core-changes/corefx/5.0/lastindexof-improved-handing-of-empty-values.md

```diff
@@ -0,0 +1,59 @@
+### LastIndexOf has improved handling of empty search strings
+
+<xref:System.String.LastIndexOf%2A?displayProperty=nameWithType> and related APIs now return correct values when searching for a zero-length (or zero-length equivalent) substring within a larger string.
+
+#### Change description
+
+In .NET Framework and .NET Core 1.0 - 3.1, <xref:System.String.LastIndexOf%2A?displayProperty=nameWithType> and related APIs might return an incorrect value when the caller searches for a zero-length substring.
+
+```csharp
+Console.WriteLine("Hello".LastIndexOf("")); // prints '4' (incorrect)
+
+ReadOnlySpan<char> span = "Hello";
+Console.WriteLine(span.LastIndexOf("")); // prints '0' (incorrect)
+```
+
+Starting with .NET 5.0, these APIs return the correct value for `LastIndexOf`.
+
+```csharp
+Console.WriteLine("Hello".LastIndexOf("")); // prints '5' (correct)
+
+ReadOnlySpan<char> span = "Hello";
+Console.WriteLine(span.LastIndexOf("")); // prints '5' (correct)
+```
+
+In these examples, `5` is the correct answer because `"Hello".Substring(5)` and `"Hello".AsSpan().Slice(5)` both produce an empty string, which is trivially equal to the empty substring that is sought.
```

Comment:
The API reference needs to reflect this.

Reply:
What do you mean by this comment? True, the sentence I included in the breaking change notification is a _drastic_ simplification of how `IndexOf` and `LastIndexOf` work, and there are some counterexamples. But as a high-level overview it's a useful way to think about the API.

See https://github.com/dotnet/runtime/blob/a057f27e860c1a59aa12e7dd9be028fba82b037f/src/libraries/System.Private.CoreLib/src/System/String.Searching.cs#L223 if you're curious as to the real logic that underlies these APIs.

Reply:
@GrabYourPitchforks I meant [this](https://docs.microsoft.com/en-us/dotnet/api/system.string.lastindexof?view=net-5.0#System_String_LastIndexOf_System_String_System_Int32_System_Int32_), and [this](https://docs.microsoft.com/en-us/dotnet/api/system.string.lastindexof?view=net-5.0#System_String_LastIndexOf_System_String_), etc..

Currently it says that the method returns the following:

> The zero-based starting index position of value if that string is found, or -1 if it is not. **If value is Empty, the return value is the last index position in this instance.**

Note the bolded part isn't correct (similar wording is there for different overloads).

Reply:
Ah, you're referring to the remarks at https://docs.microsoft.com/dotnet/api/system.string.lastindexof. I don't think those files are in this repo, so they'd not be included in this PR.

Reply:
Yep, they belong to dotnet-api-docs repo.

Reply:
I'll update the API docs now.